source/common/buffer/buffer_impl.cc

```diff
@@ -44,171 +62,574 @@ void OwnedImpl::add(const Instance& data) {
+    bool new_slice_needed = slices_.empty();
+    while (size != 0) {
+      if (new_slice_needed) {
+        slices_.emplace_front(OwnedSlice::create(size));
+      }
+      uint64_t copy_size = slices_.front()->prepend(data.data(), size);
+      size -= copy_size;
+      length_ += copy_size;
+      new_slice_needed = true;
+    }
   }
-  evbuffer_prepend(buffer_.get(), data.data(), data.size());
 }
 
 void OwnedImpl::prepend(Instance& data) {
   ASSERT(&data != this);
-  int rc =
-      evbuffer_prepend_buffer(buffer_.get(), static_cast<LibEventInstance&>(data).buffer().get());
-  ASSERT(rc == 0);
-  ASSERT(data.length() == 0);
-  static_cast<LibEventInstance&>(data).postProcess();
+  ASSERT(isSameBufferImpl(data));
+  // See the comments in move() for why we do the static_cast.
+  if (old_impl_) {
+    ASSERT(dynamic_cast<LibEventInstance*>(&data) != nullptr);
+    int rc =
+        evbuffer_prepend_buffer(buffer_.get(), static_cast<LibEventInstance&>(data).buffer().get());
+    ASSERT(rc == 0);
+    ASSERT(data.length() == 0);
+    static_cast<LibEventInstance&>(data).postProcess();
+  } else {
+    ASSERT(dynamic_cast<OwnedImpl*>(&data) != nullptr);
+    OwnedImpl& other = static_cast<OwnedImpl&>(data);
+    while (!other.slices_.empty()) {
+      uint64_t slice_size = other.slices_.back()->dataSize();
+      length_ += slice_size;
+      slices_.emplace_front(std::move(other.slices_.back()));
+      other.slices_.pop_back();
+      other.length_ -= slice_size;
+    }
+    other.postProcess();
+  }
 }
 
 void OwnedImpl::commit(RawSlice* iovecs, uint64_t num_iovecs) {
-  int rc =
-      evbuffer_commit_space(buffer_.get(), reinterpret_cast<evbuffer_iovec*>(iovecs), num_iovecs);
-  ASSERT(rc == 0);
-}
+  if (old_impl_) {
+    int rc =
+        evbuffer_commit_space(buffer_.get(), reinterpret_cast<evbuffer_iovec*>(iovecs), num_iovecs);
+    ASSERT(rc == 0);
+  } else {
+    if (num_iovecs == 0) {
+      return;
+    }
+    // Find the slices in the buffer that correspond to the iovecs:
+    // First, scan backward from the end of the buffer to find the last slice containing
+    // any content. Reservations are made from the end of the buffer, and out-of-order commits
+    // aren't supported, so any slices before this point cannot match the iovecs being committed.
+    ssize_t slice_index = static_cast<ssize_t>(slices_.size()) - 1;
+    while (slice_index >= 0 && slices_[slice_index]->dataSize() == 0) {
+      slice_index--;
+    }
+    if (slice_index < 0) {
+      // There was no slice containing any data, so rewind the iterator at the first slice.
+      slice_index = 0;
+    }
 
-void OwnedImpl::copyOut(size_t start, uint64_t size, void* data) const {
-  ASSERT(start + size <= length());
+    // Next, scan forward and attempt to match the slices against iovecs.
+    uint64_t num_slices_committed = 0;
+    while (num_slices_committed < num_iovecs) {
+      if (slices_[slice_index]->commit(iovecs[num_slices_committed])) {
```

Comment:
Right, the new implementation keeps the ASSERT because the failure to commit all of the supplied iovecs is probably a sign of a major bug. The two failure scenarios I can think of are:
* The caller did the right thing and tried to properly commit iovecs that it got from `buffer.reserve()`, but the commit didn't work due to some internal bug in the buffer code.
* Or the caller did the wrong thing and tried to `buffer.commit()` something that didn't come from that `buffer`.

Reply:
FWIW, having as many ASSERTs as possible for consistency and debug checks has some history of paying off with the buffers + fuzzers. Please keep adding!